tests/backend/test_states.py

```diff
@@ -177,20 +177,19 @@ def test_rdm(self, setup_backend, tol, cutoff, batch_size):
 
         assert np.allclose(rdm, rdm_exact, atol=tol, rtol=0)
 
-    @pytest.mark.backends("fock", "tf")
     def test_ket(self, setup_backend, pure, cutoff, batch_size, tol):
         """Test that the ket of a displaced state matches analytic result"""
         backend = setup_backend(2)
         backend.displacement(a, 0)
 
         state = backend.state()
-        if not pure:
+        if not pure and backend.short_name != "gaussian":
```

Comment:
hmm, how come? :thinking: would this not mean that the assertion after the if statement `assert state.is_pure == True` can fail?

Reply:
Ah, that's because the `pure` backend argument doesn't have any effect on the Gaussian backend.

* On the Fock backends, `pure=False` forces the backends to use the density matrix representation rather than a ket, even for pure states. It then initializes the state with the attribute `self._pure = False`.

* On the Gaussian backend, however, `pure=False` doesn't change anything --- the state is always represented as a vector of means and covariance matrix. Instead, the value of `self._pure` is set _by the state itself_, by testing the covariance matrix for purity.

This inconsistency then continues between `BaseGaussianState` and `BaseFockState`:

* `BaseFockState.is_pure` returns whether the state is represented by a density matrix or a ket, which is _not_ the same as saying whether it is pure. A pure state can always be represented as a density matrix.

* `BaseGaussianState.is_pure` return whether or not the state is pure or mixed.

These two different approaches explain why
```python
>>> eng = sf.Engine("guassian", backend_options={"pure": False})
>>> results = eng.run(prog)
>>> results.state.is_pure
True
```

There are probably two things that need to change (but out of scope for this PR):

* The `pure` backend option should probably be renamed something else? To specify that (a) it only applies to Fock backends, and (b) it only effects the internal state representation, it doesn't automatically make mix states.

* `BaseFockState.is_pure` should return whether the state is pure or mixed, and be independent of the internal representation.

Reply:
thanks, helped a lot!